### ISSUE TITLE
pyright 1.1.302 unable to solve TypeVar in some situations

### ISSUE THREAD

**Issue by stereobutter** [microsoft/pyright#4929]
**Describe the bug**
In a library I am developing I'm using a (somewhat complex) set of generic container-like classes and functions that operate on them. I noticed when writing test cases using `assert_type` that pyright 1.1.302 can fail to solve TypeVars depending on whether one uses intermediate variables or not in calls. This is a regression from pyright 1.1.301 that worked independent of how the code is structured.

**To Reproduce**
Some container-like types that use generics and overloads
```python
from typing import Generic, TypeVar, assert_type, Type, overload

T = TypeVar("T")
T1 = TypeVar("T1")
T2 = TypeVar("T2")

class Box(Generic[T]):
    def wrap(self, type: Type[T1]) -> "Box[T1]": ...

class Crate(Generic[T]):
    @overload
    def __new__(cls, box1: Box[T1]) -> "Crate[tuple[T1]]": ...
    @overload
    def __new__(cls, box1: Box[T1], box2: Box[T2]) -> "Crate[tuple[T1, T2]]": ...

Container = Box[T] | Crate[T]

def unpack(x: Container[T]) -> T: ...
```

test cases where I noticed the issue

```python
# works as expected
def example1() -> None:
    item = Crate(Box().wrap(int))  # revealed_type is Crate[tuple[int]]
    assert_type(unpack(item), tuple[int])

# same example but without intermediate variable assignments; fails using pyright 1.1.302 
def example2() -> None:
    # "assert_type" mismatch: expected "tuple[int]" but received "T@unpack" (reportGeneralTypeIssues)
    assert_type(unpack(Crate(Box().wrap(int))), tuple[int])

```

**VS Code extension or command-line**
I am running pyright (both 1.1.301 and 1.1.302) using https://pypi.org/project/pyright/ and pyright 1.1.301 in VS Code using the pylance extension.

**Comment by stereobutter** [comment]
Playing more with my API I've also found another, similar issue:

given these type definitions

```python
T = TypeVar("T")
T1 = TypeVar("T1")
T2 = TypeVar("T2")

class Box(Generic[T]):
    @overload
    def wrap(self, type: Type[T1]) -> "Box[T1]": ...
    @overload
    def wrap(self, type: None) -> "Box[None]": ...
    @overload
    def wrap(self, type: Any) -> "Box": ...

class Crate(Generic[T]):
    @overload
    def __new__(cls, box1: Box[T1]) -> "Crate[tuple[T1]]": ...
    @overload
    def __new__(cls, box1: Box[T1], box2: Box[T2]) -> "Crate[tuple[T1, T2]]": ...
    @overload
    def __new__(cls, *boxes: Box[T1]) -> "Crate[tuple[T1, ...]]": ...
    @overload
    def __new__(cls, *boxes: Box) -> "Crate[tuple]": ...
```
pyright does the following:

```python
a = Box().wrap(int)
b = Box().wrap(str)
c = Box().wrap(str)

example1 = Crate(
    reveal_type(a),  # Type of "a" is "Box[int]"
    reveal_type(b),  # Type of "b" is "Box[str]"
    reveal_type(c),  # Type of "c" is "Box[str]"
)
reveal_type(example1)  # Type of "example1" is "Crate[tuple[Unknown, ...]]"

example2 = Crate(
    reveal_type(Box().wrap(int)),  # Type of "Box().wrap(int)" is "Box[int]"
    reveal_type(Box().wrap(str)),  # Type of "Box().wrap(str)" is "Box[Unknown]"
    reveal_type(Box().wrap(str)),  # Type of "Box().wrap(str)" is "Box[Unknown]"
)

reveal_type(example2)  # Type of "example2" is "Crate[tuple[int, ...]]"
```

It appears the `def wrap(self, type: Any) -> "Box": ...` overload on `Box` does interact badly with the `def __new__(cls, *boxes: Box[T1]) -> "Crate[tuple[T1, ...]]": ...` on `Crate`.  I guess this is due to pyright somehow trying to solve the types all at once compared to the example with the variable assignments where it is forced to evaluate the arguments first?

**Comment by erictraut** [comment]
What you're doing here with the `__new__` method is highly unusual and suspect. The `cls` parameter for the `__new__` method has the type `Crate[T]`, and this would normally construct an instance of `Crate[T]`, but your `__new__` method is ignoring the incoming type and instead constructing a `Crate[tuple[T1]]`, etc. What happens if someone calls `Crate[str](Box[int]())`?

I'll see if I can make this work without breaking other more mainstream use cases.

This behavior change in pyright 1.1.302 was related to [this bug fix](https://github.com/microsoft/pyright/issues/4890).

**Comment by stereobutter** [comment]
Thank you for your response; maybe if I had used [real code](https://github.com/stereobutter/reddish) and not a toy example with the same structure things would be clearer. Could well be that this is a X-Y problem on my part and there is a better way to do things. 


In my real use case I've got a class `MultiExec` (`Crate` above) and it accepts one or more instances of `Command` ( `Box` in the example above) in its `__init__`. A command would look like this 

```python
cmd = Command("PING").into(str)  # represents the redis "PING" command and that it will return a "str"
response = redis.execute(cmd)  # pyright should infer that response is a "str"
```

This (and other commands) could be wrapped in a redis MULTI/EXEC transaction

```python
some_other_cmd = Command("...").into(int)
tx = MultiExec(cmd, some_other_cmd)
tx_response = redis.execute(tx)  # pyright should infer that tx_response is a "tuple[str, int]"
```

The `MultiExec` basically gets its specialization from the instances of `Command` it contains and it is used by `.execute()` to infer its return type.

**Comment by erictraut** [comment]
What do you expect will happen if the caller provides an explicit specialization, as in `MultiExec[int](Command(""))`? This is similar to what's happening in the failure case above because the specialization is effectively provided through bidirectional type inference. The expected type comes from the `unpack` parameter `x`.

**Comment by stereobutter** [comment]
Oh ... is the following the correct and proper way for `MultiExec` to work?

```python
class MultiExec(Generic[T]):
    @overload
    def __init__(self: "MultiExec[tuple[T1]]", c1: Command[T1]) -> None: ...
    @overload
    def __init__(
        self: "MultiExec[tuple[T1, T2]]", c1: Command[T1], c2: Command[T2]
    ) -> None: ...
```

I though that using the return type of `__new__` rather than the self type on `__init__` was the right thing to do, mainly because I remember some issue/discussion here (or on the mypy issue tracker) about the whole self-type thing not being part of any spec. I also think there was an issue here on the issue tracker where someone had issues with a self type that contained a `TypeVar`?

**Comment by erictraut** [comment]
Yes, I think the `__init__` approach is a better solution here. Not only does it avoid the error you're reporting in this bug report, but it also properly reports type mismatches if someone attempts to provide an explicit specialization that doesn't match.

**Comment by stereobutter** [comment]
What I've got now (and what looks way less scary than before; thank you so much 🙇‍♂️) is this:
```python
T = TypeVar("T")
T1 = TypeVar("T1")
T2 = TypeVar("T2")

class Command(Generic[T]):
    def __init__(self, cmd: str) -> None: ...
    def into(self: "Command[T]", type: Type[T1]) -> "Command[T1]": ...

class MultiExec(Generic[T]):
    @overload  # one arg
    def __init__(self: "MultiExec[tuple[T1]]", c1: Command[T1]) -> None: ...
    @overload  # two args
    def __init__(
        self: "MultiExec[tuple[T1, T2]]", c1: Command[T1], c2: Command[T2]
    ) -> None: ...
    @overload # more than two homogenous args
    def __init__(self: MultiExec[tuple[T, ...]], *commands: Command[T]) -> None: ...
    @overload #  more than two inhomogenous args
    def __init__(self: MultiExec[tuple], *commands: Command) -> None: ...

CommandT = Command[T] | MultiExec[T]

@overload  # one arg
def execute(cmd1: CommandT[T1]) -> tuple[T1]: ...
@overload # two args
def execute(cmd1: CommandT[T1], cmd2: CommandT[T2]) -> tuple[T1, T2]: ...
@overload  # more than two homogenous args
def execute(*cmds: CommandT[T]) -> tuple[T, ...]: ...
@overload  
def execute(*cmds: Command) -> tuple: ...
```

What still does seem to give pyright trouble is this:
```python
float_cmd = Command("...").into(float)
int_cmd = Command("...").into(int)

example1 = execute(float_cmd, int_cmd, float_cmd)  # pyright correctly infers tuple[Unknown, ...]

example2 = execute(
    Command("...").into(float),
    Command("...").into(int),
    Command("...").into(float),
)  # pyright incorrectly infers tuple[float, ...]
```

I've been toying with this for a while now and I also seem to remember that some previous version of pyright would report `tuple[int|float, ...]` for `example1` but `tuple[Unknown, ...]` is totally fine I think and the user should specialize this if they care.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.308, which I just published. It will also be included in a future release of pylance.
